File: mobile/src/main/java/com/windscribe/mobile/email/AddEmailPresenterImpl.java
/*
 * Copyright (c) 2021 Windscribe Limited.
 */

package com.windscribe.mobile.email;

import android.text.TextUtils;
import android.util.Patterns;

import com.windscribe.mobile.R;
import com.windscribe.vpn.ActivityInteractor;
import com.windscribe.vpn.api.CreateHashMap;
import com.windscribe.vpn.api.response.AddEmailResponse;
import com.windscribe.vpn.api.response.ApiErrorResponse;
import com.windscribe.vpn.api.response.GenericResponseClass;
import com.windscribe.vpn.constants.NetworkKeyConstants;

import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.HashMap;
import java.util.Map;

import javax.inject.Inject;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.observers.DisposableSingleObserver;
import io.reactivex.schedulers.Schedulers;


public class AddEmailPresenterImpl implements AddEmailPresenter {

    private final String TAG = "[add_email_p]";

    private ActivityInteractor mAddEmailInteractor;

    private AddEmailView mAddEmailView;

    private final Logger mPresenterLog = LoggerFactory.getLogger(TAG);

    @Inject
    public AddEmailPresenterImpl(AddEmailView mAddEmailView, ActivityInteractor activityInteractor) {
        this.mAddEmailView = mAddEmailView;
        this.mAddEmailInteractor = activityInteractor;
    }

    @Override
    public void onDestroy() {
        mAddEmailInteractor.getCompositeDisposable();
        if (!mAddEmailInteractor.getCompositeDisposable().isDisposed()) {
            mPresenterLog.info("Disposing network observer...");
            mAddEmailInteractor.getCompositeDisposable().dispose();
        }
        mAddEmailView = null;
        mAddEmailInteractor = null;
    }

    @Override
    public void onAddEmailClicked(String emailAddress) {
        mPresenterLog.info("Validating input email address...");
        if (TextUtils.isEmpty(emailAddress)) {
            mPresenterLog.info("Email input empty...");
            mAddEmailView.showInputError(mAddEmailInteractor.getResourceString(R.string.email_empty));
            return;
        }

        if (Patterns.EMAIL_ADDRESS.matcher(emailAddress).matches()) {
            //Post email address
            mAddEmailView.hideSoftKeyboard();
            mAddEmailView.prepareUiForApiCallStart();
            mPresenterLog.info("Posting users email address...");
            final Map<String, String> emailMap = new HashMap<>();
            emailMap.put(NetworkKeyConstants.ADD_EMAIL_KEY, emailAddress);
            emailMap.put(NetworkKeyConstants.ADD_EMAIL_FORCED_KEY, String.valueOf(1));
            mAddEmailInteractor.getCompositeDisposable().add(
                    mAddEmailInteractor.getApiCallManager()
                            .addUserEmailAddress(emailMap)
                            .subscribeOn(Schedulers.io())
                            .observeOn(AndroidSchedulers.mainThread())
                            .subscribeWith(
                                    new DisposableSingleObserver<GenericResponseClass<AddEmailResponse, ApiErrorResponse>>() {
                                        @Override
                                        public void onError(@NotNull Throwable e) {
                                            mPresenterLog
                                                    .debug("Error adding email address..." + e.getLocalizedMessage());
                                            mAddEmailView
                                                    .showToast("Sorry! We were unable to add your email address...");
                                            mAddEmailView.prepareUiForApiCallFinished();
                                        }

                                        @Override
                                        public void onSuccess(
                                                @NotNull GenericResponseClass<AddEmailResponse, ApiErrorResponse> postEmailResponseClass) {
                                            if (postEmailResponseClass.getDataClass() != null) {
                                                mAddEmailView.showToast("Added email successfully...");
                                                mPresenterLog.info("Email address added successfully...");
                                                mAddEmailView.gotoWindscribeActivity();
                                            } else {
                                                mAddEmailView.prepareUiForApiCallFinished();
                                                mAddEmailView.showToast(
                                                        postEmailResponseClass.getErrorClass().getErrorMessage());
                                                mPresenterLog.debug("Server returned error. " + postEmailResponseClass
                                                        .getErrorClass().toString());
                                                mAddEmailView.showInputError(
                                                        postEmailResponseClass.getErrorClass().getErrorMessage());
                                            }
                                        }
                                    }));


        } else {
            mAddEmailView.showInputError(mAddEmailInteractor.getResourceString(R.string.invalid_email_format));
        }
    }

    @Override
    public void setUpLayout() {
        mAddEmailView.setUpLayout(mAddEmailInteractor.getResourceString(R.string.add_email));
    }
}

File: mobile/src/main/java/com/windscribe/mobile/email/AddEmailPresenterImpl.kt
/*
 * Copyright (c) 2021 Windscribe Limited.
 */
package com.windscribe.mobile.email

import android.text.TextUtils
import android.util.Patterns
import com.windscribe.mobile.R
import com.windscribe.vpn.ActivityInteractor
import com.windscribe.vpn.api.response.AddEmailResponse
import com.windscribe.vpn.api.response.ApiErrorResponse
import com.windscribe.vpn.api.response.GenericResponseClass
import com.windscribe.vpn.constants.NetworkErrorCodes
import com.windscribe.vpn.constants.NetworkKeyConstants
import com.windscribe.vpn.repository.CallResult
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.observers.DisposableSingleObserver
import io.reactivex.schedulers.Schedulers
import org.slf4j.LoggerFactory
import javax.inject.Inject

class AddEmailPresenterImpl @Inject constructor(
    private val emailView: AddEmailView,
    private val interactor: ActivityInteractor
) : AddEmailPresenter {

    private val logger = LoggerFactory.getLogger("basic")

    override fun onDestroy() {
        if (interactor.getCompositeDisposable().isDisposed.not()) {
            interactor.getCompositeDisposable().dispose()
        }
    }

    override fun onAddEmailClicked(emailAddress: String) {
        logger.info("Validating input email address...")
        if (TextUtils.isEmpty(emailAddress)) {
            logger.info("Email input empty...")
            emailView.showInputError(interactor.getResourceString(R.string.email_empty))
            return
        }
        if (Patterns.EMAIL_ADDRESS.matcher(emailAddress).matches()) {
            //Post email address
            emailView.hideSoftKeyboard()
            emailView.prepareUiForApiCallStart()
            logger.info("Posting users email address...")
            interactor.getCompositeDisposable().add(
                interactor.getApiCallManager()
                    .addUserEmailAddress(emailAddress)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribeWith(
                        object :
                            DisposableSingleObserver<GenericResponseClass<AddEmailResponse?, ApiErrorResponse?>>() {
                            override fun onError(e: Throwable) {
                                logger
                                    .debug("Error adding email address..." + e.localizedMessage)
                                emailView
                                    .showToast("Sorry! We were unable to add your email address...")
                                emailView.prepareUiForApiCallFinished()
                            }

                            override fun onSuccess(
                                postEmailResponseClass: GenericResponseClass<AddEmailResponse?, ApiErrorResponse?>
                            ) {
                                emailView.prepareUiForApiCallFinished()
                                when (val result =
                                    postEmailResponseClass.callResult<AddEmailResponse>()) {
                                    is CallResult.Error -> {
                                        if (result.code != NetworkErrorCodes.ERROR_UNEXPECTED_API_DATA) {
                                            emailView.showToast(result.errorMessage)
                                            logger.debug(
                                                "Server returned error. " + postEmailResponseClass.errorClass.toString()
                                            )
                                            emailView.showInputError(result.errorMessage)
                                        }
                                    }
                                    is CallResult.Success -> {
                                        emailView.showToast(interactor.getResourceString(R.string.added_email_successfully))
                                        logger.info("Email address added successfully...")
                                        emailView.gotoWindscribeActivity()
                                    }
                                }
                            }
                        })
            )
        } else {
            emailView.showInputError(interactor.getResourceString(R.string.invalid_email_format))
        }
    }

    override fun setUpLayout() {
        emailView.setUpLayout(interactor.getResourceString(R.string.add_email))
    }
}